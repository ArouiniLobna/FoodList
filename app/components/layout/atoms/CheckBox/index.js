import React, { memo, useState } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

function CheckBox({ className, checked, label, dataName, onChange }) {
  const [isChecked, seIsChecked] = useState(checked);

  const handleOnChange = e => {
    onChange(e.target.getAttribute('data-name'), !isChecked);
    seIsChecked(!isChecked);
  };
  return (
    <HiddenLabel>
      <CheckboxContainer className={className}>
        <HiddenCheckbox
          data-name={dataName}
          checked={isChecked}
          onChange={handleOnChange}
        />
        <StyledCheckbox checked={isChecked}>
          <Icon viewBox="0 0 24 24">
            <polyline points="20 6 9 17 4 12" />
          </Icon>
        </StyledCheckbox>
      </CheckboxContainer>
      <span>{label}</span>
    </HiddenLabel>
  );
}

CheckBox.propTypes = {
  className: PropTypes.string, // custom class name
  checked: PropTypes.bool.isRequired,
  label: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  dataName: PropTypes.string, // the property name that needs to be updated in the json
};

const HiddenLabel = styled.label`
  width: 16px;
  height: 16px;
  line-height: 1;
  display: inline-block;
  & span {
    opacity: 0;
    position: absolute;
  }
`;
const CheckboxContainer = styled.div`
  display: inline-block;
  vertical-align: middle;
  width: 100%;
  height: 100%;
`;

const Icon = styled.svg`
  fill: none !important;
  stroke: white;
  stroke-width: 2px;
  vertical-align: unset;
`;
// Hide checkbox visually but remain accessible to screen readers.
const HiddenCheckbox = styled.input.attrs({ type: 'checkbox' })`
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
  &:focus + div {
    box-shadow: 0 0 0 3px #bfc8d9;
  }
`;

const StyledCheckbox = styled.div`
  display: inline-block;
  width: 100%;
  height: 100%;
  background: ${props => (props.checked ? '#388e3c' : '#fff')};
  border-radius: 3px;
  transition: all 150ms;
  border: ${props =>
    props.checked ? ' 1px solid #388e3c' : ' 1px solid #7d90b2'};

  ${Icon} {
    visibility: ${props => (props.checked ? 'visible' : 'hidden')};
  }
`;

export default memo(CheckBox);
