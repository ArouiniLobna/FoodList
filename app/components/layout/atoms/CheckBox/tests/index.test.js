/**
 *
 * Tests for CheckBox
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render } from 'react-testing-library';
import { mount } from 'enzyme';
// import 'jest-dom/extend-expect'; // add some helpful assertions

import CheckBox from '../index';

describe('<CheckBox />', () => {
  let container;
  let props;

  beforeEach(() => {
    props = {
      dataName: 'field-test',
      label: 'field label',
      checked: false,
      onChange: jest.fn(),
    };
    container = mount(<CheckBox {...props} />);
  });
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(<CheckBox {...props} />);
    expect(spy).not.toHaveBeenCalled();
  });

  it('should render input field with checkbox type: ', () => {
    expect(container.find('input[type="checkbox"]').length).toBe(1);
  });
  it('should render label with text: ', () => {
    expect(container.find('label span').text()).toBe('field label');
  });

  /**
   * Unskip this test to use it
   *
   * @see {@link https://jestjs.io/docs/en/api#testskipname-fn}
   */
  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(<CheckBox {...props} />);
    expect(firstChild).toMatchSnapshot();
  });
});
