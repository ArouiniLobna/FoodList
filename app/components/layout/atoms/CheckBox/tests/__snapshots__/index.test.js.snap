// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckBox /> Should render and match the snapshot 1`] = `
<label
  class="CheckBox__HiddenLabel-sc-3ds2-0 gnQLZV"
>
  <div
    class="CheckBox__CheckboxContainer-sc-3ds2-1 hgbZRk"
  >
    <input
      class="CheckBox__HiddenCheckbox-sc-3ds2-3 chOmlZ"
      data-name="field-test"
      type="checkbox"
    />
    <div
      class="CheckBox__StyledCheckbox-sc-3ds2-4 eVKOMr"
    >
      <svg
        class="CheckBox__Icon-sc-3ds2-2 dpDFqD"
        viewBox="0 0 24 24"
      >
        <polyline
          points="20 6 9 17 4 12"
        />
      </svg>
    </div>
  </div>
  <span>
    field label
  </span>
</label>
`;
